### ISSUE TITLE
loss is not stable

### ISSUE THREAD

**Issue by HariKrishna-Vydana** [mkocabas/focal-loss-keras#5]
the loss turns to NAN after an epoch. i have added K.epsilon() inside the log that was fine if gama is 2,1 if gama less than 1 thena the loss turns to NAN again

**Comment by MoyanZitto** [comment]
same problem, nan when training.

**Comment by hushunda** [comment]
very very not stable ，it is easy to turns to Nan

**Comment by manyu90** [comment]
Add in  clipping to y_pred  to be between (eps,1-eps)

**Comment by bonlime** [comment]
It looks like bad initialization is the problem. What works for me is to run 2 epochs with binary_crossentropy from keras.losses, and then compile model again, using focal loss and train until convergence.

**Comment by cetingzhou** [comment]
You can try alpha=0.25. This make my training much more stable.

**Comment by bonlime** [comment]
I've tried adding clipping as suggested above and using alpha=0.25, but sometimes it still fails. 
Why default value for alpha is 2? it makes no sense, because it should be less than 1 to weight values properly

**Comment by mkocabas** [comment]
I've set the `alpha=.25` as in the paper.

**Comment by sarim-zafar** [comment]
I am still getting nan after a few epochs even with alpha=0.25 and i even added k.epsilon() in both logs.
Any fix recommended?

**Comment by bonlime** [comment]
@sarim-zafar I tried to find what causes this nans, but didn't succeed, maybe you can find it out

**Comment by toosyou** [comment]
I think it's because the calculation of cross_entropy is not stable.
I am using 
```python3
def focal_loss_fixed(y_true, y_pred):
        y_true = K.flatten(y_true)
        y_pred = K.flatten(y_pred)

        alpha_factor = keras.backend.ones_like(y_true) * alpha
        alpha_factor = tf.where(keras.backend.equal(y_true, 1), alpha_factor, 1 - alpha_factor)
        focal_weight = tf.where(keras.backend.equal(y_true, 1), 1 - y_pred, y_pred)
        focal_weight = alpha_factor * focal_weight ** gamma

        cls_loss = focal_weight * K.binary_crossentropy(y_true, y_pred)
        return  K.sum(cls_loss)
```
then it's working fine.

**Comment by braindancer** [comment]
@toosyou I believe yours is not the focal loss implemention since binary_crossentropy is already a scalar

**Comment by zheng-xing** [comment]
@braindancer  Agreed! @toosyou 's implementation is wrong. The focal weights should be adjusted pixel by pixel given each pixel's prediction and ground truth labels.

In summary, to make it more stable, one should:
1) Train the model use cross entropy loss for a few epochs first;
2) Add "y_pred = K.clip(y_pred, eps, 1-eps)".
